### ISSUE TITLE
LH accepts function with malformed spec

### ISSUE THREAD

**Issue by ranjitjhala** [ucsd-progsys/liquidhaskell#292]
See https://github.com/ucsd-progsys/liquidhaskell/blob/master/tests/todo/badspec.hs
See https://github.com/ucsd-progsys/liquidhaskell/blob/master/tests/todo/badspec1.hs

This should be rejected...

**Comment by ranjitjhala** [comment]
Another variant. The actual bugs are different of course, in the latter case I think we probably just take a prefix of the args and drop the rest which is why `badspec1.hs` is safe. 

Silly but very severe bug, we shouldn't be **accepting** programs with malformed specs!

**Comment by nikivazou** [comment]
I think @spinda (Michael) has a fix for the second variant (not merged yet.)

**Comment by spinda** [comment]
I believe the cause for the second variant is the same as for #287.

**Comment by spinda** [comment]
~~I believe the cause for the second variant is the same as for #287.~~

Looking at it again, that looks like a totally different issue. The problem is that it's accepting `Foo 1000` instead of `Fooz 1000`, right?

**Comment by ranjitjhala** [comment]
Yes the 1000 is somehow ignored in the application...

**Comment by spinda** [comment]
Just as an update, I'm working on fixing this right now.

**Comment by ranjitjhala** [comment]
Folks, what is going on with this? 

We need to fix it ASAP, because its a rather horrible and urgent bug. 

I have added another version, this time in `tests/neg` to remind ourselves 
that the build is broken until this is fixed.

**Comment by spinda** [comment]
Ack, ack, ack. Right, this is the plugHoles issue.

**Comment by ranjitjhala** [comment]
Is it the case that plug holes is going to take a while to fix?

**Comment by spinda** [comment]
No, I believe it can be fixed on its own. I'll work on it asap if no one
else has gotten to it before then.

**Comment by spinda** [comment]
Working on this right now. (Well, first I'm waiting for my system to update so that I can compile the latest `fixpoint`...)

**Comment by spinda** [comment]
For `badspec.hs`, `plugHoles` turns

```haskell
module Goo where

{-@ foo :: xs:_ -> {v:_ | this = rubbish } @-}
foo _ _ = 0
```

into

```haskell
module Goo where

{-@ foo :: Num a => xs:t -> {v:(t -> a) | this = rubbish } @-}
foo :: Num a => t -> t -> a
foo _ _ = 0
```

and then LiquidHaskell seems to ignore refts on `(->)`, as that passes on its own.

A separate issue pops up if I leave off `foo :: Num a => t -> t -> a`. `liquidhaskell` complains that the LiquidHaskell type `foo :: Num a => t -> t -> a` does not match the inferred Haskell type `foo :: Num a => t -> t -> a`&mdash;so the type is not equal to itself.

**Comment by spinda** [comment]
`badspec{1,2}.hs` come from a different problem: GHC's `expandTypeSynonyms` doesn't raise an error if a type synonym has invalid arguments (as it's just `Type -> Type`, no monad). This why the issue only occurs for Haskell type synonyms and not LiquidHaskell type synonyms:

```haskell
module Zoo where

type Foo = Int

{-@ type Fooz X = {v:Int | X < v} @-}

{-@ bob :: Foo 1000 @-}
bob = 10 :: Int
```

which is "SAFE", vs.

```haskell
module Zoo where

type Foo = Int

{-@ type Fooz = {v:Int | 1 < v} @-}

{-@ bob :: Fooz 1000 @-}
bob = 10 :: Int
```

which produces a LiquidHaskell error.

There must be a function somewhere in the GHC API that can raise these errors for us. I'd rather track one down and use that than have us implement these checks ourselves...

**Comment by spinda** [comment]
The relevant GHC code for that is here: https://github.com/ghc/ghc/blob/ghc-7.8/compiler/types/TyCon.lhs#L1427

GHC silently discards (doesn't expand) type synonyms with too many arguments. Odd behavior, but I'm sure there's a reason for it within GHC.

**Comment by ranjitjhala** [comment]
Sorry I don't quite understand. Does it complain now or not?

On Fri, Mar 27, 2015 at 4:37 PM, Michael Smith <notifications@github.com>
wrote:

> For badspec.hs, plugHoles turns
>
> module Goo where
> {-@ foo :: xs:_ -> {v:_ | this = rubbish } @-}
> foo _ _ = 0
>
> into
>
> module Goo where
> {-@ foo :: Num a => xs:t -> {v:(t -> a) | this = rubbish } @-}foo :: Num a => t -> t -> a
> foo _ _ = 0
>
> and then LiquidHaskell seems to ignore refts on (->), as that passes on
> its own.
>
> A separate issue pops up if I leave off foo :: Num a => t -> t -> a.
> liquidhaskell complains that the LiquidHaskell type foo :: Num a => t ->
> t -> a does not match the inferred Haskell type foo :: Num a => t -> t ->
> a--so the type is not equal to itself.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/292#issuecomment-87121461>
> .
>



-- 
Ranjit.

**Comment by spinda** [comment]
Breaking it down:

 1. `badspec.hs`: LiquidHaskell ignores all predicates on `(->)`, so `this == rubbish` isn't checked or used at all as the hole gets filled with `(t -> a)`. It doesn't complain on the test case (when it should), but it does complain on
    ```haskell
    module Goo where
    
    {-@ foo :: xs:_ -> {v:_ | this = rubbish } @-}
    foo _ = 0
    ```
    in which case `v:_` becomes `v:a`.
 1. `badspec{1,2}.hs`: We use `expandTypeSynonyms` to handle Haskell type synonym expansion, but this silently passes through cases where a type synonym is applied to too many arguments. LiquidHaskell should complain here, but it doesn't. It *does* complain (as it should) when we're using a LiquidHaskell synonym (like `{-@ type Nat = {v:Int | 0 <= v} @-}`), as we do our own checks in that case.

**Comment by spinda** [comment]
I'm currently trying to find a GHC function to do these sanity checks on `Type`s. If anyone knows one, please let me know! If I can't find one, I'll fall back to implementing one myself...

**Comment by ranjitjhala** [comment]
Can we just compare the number of params with that expected by the synonym in ghc?

**Comment by spinda** [comment]
Yes, but I'd like to use existing GHC mechanisms where we can... I think `tcLHsType` will do the job, but otherwise I'll just do that.

**Comment by spinda** [comment]
Ah, `tcLHsType` can only be called from within `TcM`... Alright, I'll just add the check in myself and let this get sorted out by the GSoC work.

**Comment by gridaphobe** [comment]
(1) should be easy to fix, just throw an error if a function is refined with anything other than TRUE.

for (2), the issue may be that Haskell type synonyms can be eta reduced just like functions, e.g. a common pattern is

```haskell
type MyMonad = State Foo
```

where the "result" type `a` is left out. Thus when GHC sees `MyMonad Int` it doesn't know immediately whether the type is well-formed (`State` could be another alias). The check might need to be deferred until the last moment.

**Comment by spinda** [comment]
Ah, I see. That makes sense for (2), thanks. For (1), we don't support refinements on functions at all then, right?

**Comment by ranjitjhala** [comment]
I don't buy the argument re 2 -- in this case the parameter we are passing in is a binder n and not a type variable, so we cannot blame GHc

**Comment by ranjitjhala** [comment]
Yes, I agree with Michael's plan

**Comment by gridaphobe** [comment]
@ranjitjhala oh I agree completely that it's our fault, I was just giving a possible explanation for why `expandTypeSynonyms` doesn't throw any errors.

**Comment by spinda** [comment]
@ranjitjhala, by "2" do you mean `badspec2.hs`? I meant to refer to both `badspec{1,2}.hs` there.

For `badspec2.hs`, `Point` is applied to too many type synonyms, and since `expandTypeSynonyms` doesn't check for that we need to add our own check.

If we remove this issue, then we do actually error out due to the overlapping binder and type variable names, but in a weird way:

```haskell
LiquidHaskell Copyright 2009-14 Regents of the University of California. All Rights Reserved.


**** DONE:  annotate ***********************************************************
 

**** ERROR:  *******************************************************************


 tests/crash/badspec2.hs:6:5: Error: Specified Type Does Not Refine Haskell Type for Foo.foo
 Haskell: GHC.Types.Int -> [GHC.Types.Double]
 Liquid : forall n. GHC.Types.Int -> [n]
```

Do we do any checks for this already?

**Comment by gridaphobe** [comment]
Hrm, I'm not sure what you did to "remove" the issue, but that error is completely wrong. There shouldn't be a `forall` in the liquid type.

**Comment by ranjitjhala** [comment]
The "n" has nothing to do with expand type synonym thing here, because the n is a binder NOT a haskell tyvar it should not even go to ghc...

**Comment by spinda** [comment]
"remove" the too-many-arguments issue by adding a parameter to `Point`, isolating the "binder as type variable" problem.

That error message is actually because I left in the explicit signature for `foo`... If I remove it, we get:

```haskell
module Foo where

type List a = [a]
type Point a = List a

{-@ foo :: n:Nat -> Point n @-}
foo :: Int -> List n
foo _ = []
```
```
LiquidHaskell Copyright 2009-14 Regents of the University of California. All Rights Reserved.


**** DONE:  Extracted Core using GHC *******************************************
 

**** DONE:  generateConstraints ************************************************
 

**** START: fixpoint ***********************************************************
 


========================================================
© Copyright 2009 Regents of the University of California.
All Rights Reserved.
========================================================
.|
SAT

**** DONE:  fixpoint ***********************************************************
 

**** DONE:  solve **************************************************************
 

**** DONE:  annotate ***********************************************************
 

**** SAFE **********************************************************************
```

I'm not sure what the correct way to handle this is. The current behavior (if I understand correctly) is for type variables to exist in their own namespace, and overlapping with binder names is allowed. Should we forbid this?

**Comment by spinda** [comment]
@ranjitjhala, it does in the original test case:

```haskell
module Foo where

type List a = [a]
type Point  = List Double

{-@ foo :: n:Nat -> Point n @-}
foo :: Int -> List Double
foo _ = []
```

which combined two separate issues: allowing "Point" to be applied to too many type arguments, and using a binder as a type variable.

**Comment by ranjitjhala** [comment]
I have no idea what's going on :)

Bottom line : "n" is not something we can ask ghc to expand because it's a binder and that is a bug that we should catch.

**Comment by ranjitjhala** [comment]
> The current behavior (if I understand correctly) is for type variables to exist in their own namespace, and overlapping with binder names is allowed. Should we forbid this?

Absolutely this should be completely disallowed it makes no sense!

**Comment by ranjitjhala** [comment]
This overlapping is exactly the cause of the bug yeah, we should check for it and project such programs.

**Comment by spinda** [comment]
Right, okay, sorry. :) Let me make I have everything straight now:

So currently we (accidentally) treat binders and type variables like GHC treats type constructors and data constructors: they exist in different namespaces and their names can overlap. There's no ambiguous case (I think) where something can either be a binder or a type variable.

The signature

```haskell
{-@ foo :: n:Nat -> Point n @-}
```

is equivalent to

```haskell
{-@ foo :: forall n. n:Nat -> Point n @-}
```

is equivalent to

```haskell
{-@ foo :: forall n0. n1:Nat -> Point n0 @-}
```

because `Point n` must always be referring to the type variable `n`, not the binder `n`.

**Instead**, we want to add a check for this case and reject programs that use this, because this is confusing behavior. I can do that!

**Comment by gridaphobe** [comment]
It might be easiest to separate binders and tyvars immediately after parsing. Then
```
{-@ foo :: n:Nat -> Point n @-}
```
would treat both occurences of `n` as binders, and throw an error when we apply the Haskell alias `Point` to the binder `n`.

**Comment by spinda** [comment]
Working on `badspec.hs`, is there a reason why `RFun` has an `rt_reft` field at all? Is it used anywhere, or can it be taken out?

**Comment by ranjitjhala** [comment]
It's there for some very technical but important reasons -- can you keep it in? We always want to make sure it's just true ie "top"

**Comment by spinda** [comment]
Sure, I can keep it in. Somewhere between `makePluggedSigs` and `checkRType`, `rt_reft` for all `RFun` is being set to TRUE. Going to track that down next.

**Comment by spinda** [comment]
The refinements for `RFun` were being lost between translations to and from `RTypeRep`. I've fixed this by storing them in `RTypeRep` in the same way that we store binders.

**Comment by nikivazou** [comment]
No need to store them. The refinements on functions are always true.

**Comment by spinda** [comment]
We need to store them in order to report errors if any of them are not true.

**Comment by ranjitjhala** [comment]
But remember that soundness issue involving instantiating tyvarS (with refinements) with functions (without refinements)...

**Comment by nikivazou** [comment]
Yes this is why we added refinements to functions. But they are always
true, right?

**Comment by spinda** [comment]
Right now a user can specify a refinement on a function type, and it will
be silently discarded and replaced with true by the end of the Bare
processing. This is why badspec.hs is marked as "SAFE", which I'm fixing
with this.

**Comment by ranjitjhala** [comment]
Yup!

**Comment by spinda** [comment]
They *should* always be true, but they can end up not being true, and we
need to handle that instead of ignoring bad refinements.

**Comment by nikivazou** [comment]
I see.
But then, the parser should reject refining function type.

**Comment by spinda** [comment]
The function type have a refinement attached to it after parsing, through
type synonym expansion or through plugHoles (as in badspec.hs). Also,
reporting it as a parser means the user sees fewer possible errors per
LiquidHaskell run, vs. reporting it in checkGhcSpec.

**Comment by nikivazou** [comment]
As Ranjit says for soundness function types should have refinements.

What I am saying is that since the user cannot specify non trivial function
refinements the syntax should not allow refinements on functions.
I believe this is simpler than specifying that the syntax allows these
refinements, but they should be true.

If the syntax does not allow refinements on functions and the user writes
them then we have a parsing error, and the program fails without analyzing
the rest of your code, as GHC does.

**Comment by spinda** [comment]
That doesn't help with the original issue, though (badspec.hs). An invalid
refinement is assigned to a function type via plugHoles, and by discarding
the refinement instead of passing it through (to either check or Fixpoint),
we mark a malformed spec as "SAFE".

GHC actually allows some things through the parsing stage so that it can
report more errors at once, and to avoid overcomplicating the parsing code.
But we could make this a parse error or something, in addition to passing
refinements all the way through, if we only want to eliminate explicit
refinements on functions. That would create an inconcistency when using
plugHoles, though.

**Comment by gridaphobe** [comment]
@nikivazou the parser already forbids writing types like `{v:(a -> b) | p}`, the problem is that you can instead write `{v:_ | p}` and have the hole instantiated to a function type, which the parser **cannot** prevent.

**Comment by spinda** [comment]
@gridaphobe actually the parser doesn't forbid that currently.

```haskell
module Goo where

{-@ foo :: Num a => xs:t -> {v:(t -> a) | this = rubbish } @-}
foo :: Num a => t -> t -> a
foo _ _ = 0
```

```
LiquidHaskell Copyright 2009-14 Regents of the University of California. All Rights Reserved.


**** DONE:  Extracted Core using GHC *******************************************
 

**** DONE:  generateConstraints ************************************************
 

**** START: fixpoint ***********************************************************
 


========================================================
© Copyright 2009 Regents of the University of California.
All Rights Reserved.
========================================================
.|
SAT

**** DONE:  fixpoint ***********************************************************
 

**** DONE:  solve **************************************************************
 

**** DONE:  annotate ***********************************************************
 

**** SAFE **********************************************************************
```

**Comment by spinda** [comment]
@gridaphobe But yes, that's exactly the problem with only checking it in the parser.

**Comment by gridaphobe** [comment]
Well that's what I get for trying to read the parser code instead of just running an example.. In that case, I don't see a reason to tweak the parser, better to catch it in Bare along with all the other errors we look for.

**Comment by spinda** [comment]
And that's what I have implemented right now. I've modified the code to pass through refinements on function types all the way through the Bare pipeline, and then I added a check to checkGhcSpec to report errors. I should have it up in a PR soon-ish.

**Comment by nikivazou** [comment]
@spinda, you fixed this, right?

**Comment by spinda** [comment]
Yes, this is fixed in #368 and #369. Forgot to close it.

**Comment by nikivazou** [comment]
Thanks!

On Fri, Apr 17, 2015 at 3:45 PM, Michael Smith <notifications@github.com>
wrote:

> Closed #292 <https://github.com/ucsd-progsys/liquidhaskell/issues/292>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/292#event-284300392>
> .
>



-- 
Niki Vazou

**Comment by ranjitjhala** [comment]
Thanks!!
